### PR TITLE
A revised querying section

### PR DESCRIPTION
**Pull request by martinthomson**
This explainer still leaves a bit to be desired when it comes to introductory material, but this might help.

### CODE REVIEW

File: IPA-End-to-End.md [hunk 1]
```diff
@@ -120,15 +120,28 @@ These events will occur on different _websites/apps_, so they then need to be ga
...
 
-When making a _fanout query_, the _report collector_ uses the information it receives about events to decide whether to include an event in the query and what value is attached to the event.  The _report collector_ also assigns a _breakdown key_ to each _source event_ and a _trigger value_ to each _trigger value_.  The _report collector_ also assigns a value of its choice for the _attribution constraint ID_ of all events.
+* A _trigger fanout query_ helps a site that buys ads to understand how its advertising is helping to drive outcomes on its _website/app_.  A _trigger fanout query_ can only contain _trigger reports_ from a single _source website/app_.  A _trigger fanout query_ includes _source reports_ from multiple sites, where each report might represent an ad impression or click.
+
+The _report collector_ assigns a number of values to _source reports_ and _trigger reports_ when making a query.  The values that are assigned to each report determine what the query results mean:
```
Comment (eriktaubeneck) [inline IPA-End-to-End.md:131]:
```suggestion
Before making a query, _source websites/apps_ and _trigger websites/apps_ respectively generate _source reports_ and _trigger_reports_ by associating _encrypted match keys_ with other values. A collection of _source reports_ and _trigger reports_ with associated values determine what a query result means:
```

File: IPA-End-to-End.md [hunk 2]
```diff
@@ -120,15 +120,28 @@ These events will occur on different _websites/apps_, so they then need to be ga
...
+
+* A _breakdown key_ is associated with each _source report_ in a query.  The _breakdown key_ is used to split _source reports_ into groups for aggregation.  A _breakdown key_ can represent an advertising campaign, an advertising channel, a set of creatives, or any combination of these with other data from the event.  The output of the query includes an aggregate value for each _breakdown key_.
+
+* A _trigger value_ is associated with each _trigger report_.  If the _trigger report_ is matched with a _source reports_ by the MPC, the _trigger value_ contributes to the aggregated value for the _breakdown key_ from the _source report_.  Note that [sensitivity capping](#user-level-sensitivity-capping) might mean that some _trigger values_ do not contribute to the final result.
```
Comment (eriktaubeneck) [inline IPA-End-to-End.md:135]:
```suggestion
* A _trigger value_ is associated with each _trigger report_.  If the _trigger report_ is matched with a _source reports_ by the MPC, the _trigger value_ contributes to the aggregated value for the _breakdown key_ from the _source report_.  Note that [sensitivity capping](#user-level-sensitivity-capping) might mean that some _trigger values_ do not contribute to the final result.

* A _timestamp_ is associated with each _source report_ and each _trigger report_. These timestamps are used to validate that we only attribute events when the _trigger report timestamp_ is after the _source report timestamp_. We are likely to also consider more sophisticated operations, such as checking that timestamps are within a fixed relative distance, e.g., a day or a week.

```

File: IPA-End-to-End.md [hunk 3]
```diff
@@ -120,15 +120,28 @@ These events will occur on different _websites/apps_, so they then need to be ga
...
+
+* A _attribution constraint ID_ can be added to both _source reports_ and _trigger reports_.  Reports with different _attribution constraint IDs_ will not be matched to each other, which allows a site to include _source reports_ and _trigger reports_ that it does not wish to have attributed together.  For instance, if the goal is to perform attribution for distinct product segments, assigning a different _attribution constraint ID_ to reports from each distinct segment will ensure that impressions from one segment is not attributed to conversions from the other segment.
```
Comment (eriktaubeneck) [inline IPA-End-to-End.md:137]:
```suggestion
* A _attribution constraint ID_ can be associated to both _source reports_ and _trigger reports_.  Reports with different _attribution constraint IDs_ will not be matched to each other, which allows a site to include _source reports_ and _trigger reports_ that it does not wish to have attributed together.  For instance, if the goal is to perform attribution for distinct product segments, assigning a different _attribution constraint ID_ to reports from each distinct segment will ensure that impressions from one segment is not attributed to conversions from the other segment.
```

Comment (eriktaubeneck) [reply IPA-End-to-End.md:137]: should we continue to use this verb?

File: IPA-End-to-End.md (hunk 2 above)
Comment (martinthomson) [reply IPA-End-to-End.md:135]: Maybe just:

> * A _timestamp_ for each report.

I don't think that we need to go into details about validation and whatnot just yet.

File: IPA-End-to-End.md (hunk 3 above)
Comment (martinthomson) [reply IPA-End-to-End.md:137]:
```suggestion
* A _attribution constraint ID_ can be associated with both _source reports_ and _trigger reports_.  Reports with different _attribution constraint IDs_ will not be matched to each other, which allows a site to include _source reports_ and _trigger reports_ that it does not wish to have attributed together.  For instance, if the goal is to perform attribution for distinct product segments, assigning a different _attribution constraint ID_ to reports from each distinct segment will ensure that impressions from one segment is not attributed to conversions from the other segment.
```

File: IPA-End-to-End.md [hunk 4]
```diff
@@ -120,15 +120,28 @@ These events will occur on different _websites/apps_, so they then need to be ga
...
+
+The _report collector_ uses the information it receives about events to decide whether to include an event in the query and what values to attach to the event.  Choosing which events are included and the values that are associated with each gives the _report collector_ the ability to make different queries from the same set of events, subject only to differential privacy constraints on those queries.
```
Comment (eriktaubeneck) [inline IPA-End-to-End.md:139]:
```suggestion
In some cases, the _report collector_ may receive this associated information directly, using it to construct _source reports_ and _trigger reports_, and construct queries with different subsets of _reports_.  Choosing which _reports_ are included and the values that are associated with each gives the _report collector_ the ability to make different queries from the same set of events, subject only to differential privacy constraints on those queries.
```

File: IPA-End-to-End.md (hunk 2 above)
Comment (eriktaubeneck) [reply IPA-End-to-End.md:135]: Sure, that's fine. I was mostly just following the form of the others 🙂

File: IPA-End-to-End.md (hunk 3 above)
Comment (eriktaubeneck) [reply IPA-End-to-End.md:137]: ha yes, if we use the same verb we should use it correctly!

**Comment by eriktaubeneck** [comment]
A few suggestions but LGTM

File: IPA-End-to-End.md (hunk 4 above)
Comment (martinthomson) [reply IPA-End-to-End.md:139]: Hmm, I think that this is potentially quite confusing.

I think that what you are trying to explain is that there are two models (with a mix also being possible):

* where the report collector assembles the query based on more or less full information from the advertisers and publishers
* where advertisers or publishers generate their portion of the query and the report collector only interleaves those parts

Is that right?  Because I don't think that the latter is really possible in practice and talking about it is only going to confuse things.

**Review by eriktaubeneck — APPROVED** [review]

File: IPA-End-to-End.md (hunk 4 above)
Comment (eriktaubeneck) [reply IPA-End-to-End.md:139]: Just trying to line this up with above. The two cases I'm thinking of:
1. advertisers and publishers generate their own reports, and send them to the report collector
2. report collector is running JS on their sites and generates those reports themselves

maybe these aren't worth distinguishing? 

As for the two models you point out, I do see two distinct versions:
1. advertisers and publishers send events (in some form) to a report collector
2. a large publisher or large advertiser has all partner advertisers/publishers (respectively) send them their events directly, and they act as their own report collector.

this also maybe isn't worth distinguishing?

Comment (martinthomson) [reply IPA-End-to-End.md:139]: Maybe we can add some discussion about these to the section on gathering events.  Is a follow-up OK for you?

Comment (eriktaubeneck) [reply IPA-End-to-End.md:139]:
```suggestion
The _report collector_ uses these associated values on events it receives to construct queries.  Choosing which events are included or adjusting the values that are associated with each gives the _report collector_ the ability to make different queries from the same set of events, subject only to differential privacy constraints on those queries.
```

how about this?

File: IPA-End-to-End.md [hunk 5]
```diff
@@ -120,15 +120,28 @@ These events will occur on different _websites/apps_, so they then need to be ga
...
 
-When making a _fanout query_, the _report collector_ uses the information it receives about events to decide whether to include an event in the query and what value is attached to the event.  The _report collector_ also assigns a _breakdown key_ to each _source event_ and a _trigger value_ to each _trigger value_.  The _report collector_ also assigns a value of its choice for the _attribution constraint ID_ of all events.
+* A _trigger fanout query_ helps a site that buys ads to understand how its advertising is helping to drive outcomes on its _website/app_.  A _trigger fanout query_ can only contain _trigger reports_ from a single _source website/app_.  A _trigger fanout query_ includes _source reports_ from multiple sites, where each report might represent an ad impression or click.
+
+Before making a query, _source websites/apps_ and _trigger websites/apps_ respectively generate _source reports_ and _trigger_reports_ by associating _encrypted match keys_ with other values. A collection of _source reports_ and _trigger reports_ with associated values determine what a query result means:
```
Comment (eriktaubeneck) [inline IPA-End-to-End.md:131]:
```suggestion
Before making a query, _source websites/apps_ and _trigger websites/apps_ provide a _report collector_ with additional information, enabling it to generate _source reports_ and _trigger_reports_ by associating _encrypted match keys_ with other values. A collection of _source reports_ and _trigger reports_ with associated values determine what a query result means:
```

Comment (eriktaubeneck) [reply IPA-End-to-End.md:131]: Ok, I think this helps clean it up a bit.

File: IPA-End-to-End.md [hunk 6]
```diff
@@ -120,15 +120,28 @@ These events will occur on different _websites/apps_, so they then need to be ga
...
+
+* A _attribution constraint ID_ can be associated with both _source reports_ and _trigger reports_.  Reports with different _attribution constraint IDs_ will not be matched to each other, which allows a site to include _source reports_ and _trigger reports_ that it does not wish to have attributed together.  For instance, if the goal is to perform attribution for distinct product segments, assigning a different _attribution constraint ID_ to reports from each distinct segment will ensure that impressions from one segment is not attributed to conversions from the other segment.
+
+The _report collector_ uses these associated values on events it receives to construct queries.  Choosing which events are included or adjusting the values that are associated with each gives the _report collector_ the ability to make different queries from the same set of events, subject only to differential privacy constraints on those queries.
```
Comment (eriktaubeneck) [inline IPA-End-to-End.md:139]:
```suggestion
The _report collector_ uses these associated values on events it receives to construct queries.  Choosing which events are included and the values that are associated with each gives the _report collector_ the ability to make different queries from the same set of events, subject only to differential privacy constraints on those queries.
```
with the change above, we can revert this change
